### PR TITLE
[harmony #2057] Enable Atto precision transactions using the CLI

### PR DESCRIPTION
**Pull request by janet-harmony**
* Replace big.Int with numeric.Dec to enable enough precision for Atto transactions
* Check transaction amount + gas price when verifying existing balance before transactions
* Enable scientific notation input
* Enable user to input desired gas-price & gas-limit
* Set default gas to 1 Gwei
* Calculate gas-price based on gas-limit

Staking transaction is currently untested.

### CODE REVIEW

File: pkg/common/numeric.go [hunk 1]
```diff
@@ -2,10 +2,44 @@ package common
...
 
 	"github.com/harmony-one/harmony/common/denominations"
+	"github.com/harmony-one/harmony/numeric"
 )
 
 func NormalizeAmount(value *big.Int) *big.Int {
 	return value.Div(value, big.NewInt(denominations.Nano))
 }
+
+func Pow(base numeric.Dec, exp int) numeric.Dec {
```
Comment (fxfactorial) [inline pkg/common/numeric.go]: did we really need to implement this?

File: pkg/common/numeric.go [hunk 2]
```diff
@@ -2,10 +2,44 @@ package common
...
+		if (exp == 0) {
+			break
+		}
+		base = base.Mul(base)
+	}
+	return result
+}
+
+func NewDecFromString (i string) (numeric.Dec, error) {
+	pattern, _ := regexp.Compile("[0-9]+\\.{0,1}[0-9]*e-{0,1}[0-9]+")
```
Comment (fxfactorial) [inline pkg/common/numeric.go]: make the regex be package scope, no need to make over and over

File: pkg/transaction/controller.go [hunk 3]
```diff
@@ -96,16 +100,15 @@ func (C *Controller) verifyBalance(amount float64) {
...
 	currentBalance, _ := balanceRPCReply["result"].(string)
-	balance, _ := big.NewInt(0).SetString(currentBalance[2:], 16)
-	balance = common.NormalizeAmount(balance)
-	transfer := big.NewInt(int64(amount * denominations.Nano))
-
-	tns := float64(transfer.Uint64()) / denominations.Nano
-	bln := float64(balance.Uint64()) / denominations.Nano
+	bal, _ := new(big.Int).SetString(currentBalance[2:], 16)
+	balance := numeric.NewDecFromBigInt(bal)
+	// Add transaction amount and gas price in Atto
```
Comment (fxfactorial) [inline pkg/transaction/controller.go]: is this a todo or a node? prefix comment with // NOTE

File: pkg/transaction/controller.go [hunk 4]
```diff
@@ -96,16 +100,15 @@ func (C *Controller) verifyBalance(amount float64) {
...
+	gasAsDec := numeric.NewDecFromBigInt(big.NewInt(int64(C.transactionForRPC.params["gas-price"].(uint64))))
+	total := C.transactionForRPC.params["transfer-amount"].(numeric.Dec).Add(gasAsDec)
 
-	if tns > bln {
+	if total.GT(balance) {
```
Comment (fxfactorial) [inline pkg/transaction/controller.go]: maybe GTE?

File: pkg/transaction/controller.go [hunk 5]
```diff
@@ -123,51 +126,42 @@ func (C *Controller) sendSignedTx() {
...
-	inputData, _ := base64.StdEncoding.DecodeString(rawInput)
-	// NOTE Need to add more intrisicGas
-	// in order to include more data in inputData
-	gas, _ := core.IntrinsicGas(inputData, false, true)
-	C.transactionForRPC.params["gas"] = gas
+	C.transactionForRPC.params["gas-limit"] = uint64(gasLimit)
 }
 
-func (C *Controller) setGasPrice() {
+func (C *Controller) setGasPrice(gasPrice int) {
```
Comment (fxfactorial) [inline pkg/transaction/controller.go]: make gas price be uint64 as parameter

File: pkg/transaction/controller.go (hunk 4 above)
Comment (janet-harmony) [reply pkg/transaction/controller.go]: Transaction total cost == balance is valid. Your balance should be exactly 0 after.

File: pkg/common/numeric.go (hunk 1 above)
Comment (janet-harmony) [reply pkg/common/numeric.go]: Yes

**Pull request merged**